cpp/src/stream_compaction/copy_if.cuh

```diff
@@ -326,51 +343,68 @@ gdf_column copy_if(gdf_column const &input, Filter filter,
-
-    void *data = nullptr;
-    gdf_valid_type *valid = nullptr;
-    RMM_ALLOC(&data, output_size * column_byte_width, stream);
+  table output;
 
-    if (input.valid != nullptr) {
-      auto bytes = gdf_valid_allocation_size(output_size);
-      RMM_ALLOC(&valid, bytes, stream);
-      CUDA_TRY(cudaMemsetAsync(valid, 0, bytes, stream));
-    }
-
-    CUDF_EXPECTS(GDF_SUCCESS == gdf_column_view(&output, data, valid,
-                                                output_size, input.dtype),
-                "cudf::apply_boolean_mask failed to create output column view");
+  if (output_size == input.num_rows()) {
+    output = cudf::copy(input);
+  }
+  else if (output_size > 0) {
+    // Allocate/initialize output columns
+    output = cudf::allocate_like(input, output_size, true, stream);
 
     // 4. Scatter the output data and valid mask
-    cudf::type_dispatcher(output.dtype,
-                          scatter_functor<Filter, block_size, per_thread>{},
-                          output, input, block_offsets, filter,
-                          input.valid != nullptr, stream);
-
-    CHECK_STREAM(stream);
+    for (int col = 0; col < input.num_columns(); col++) {
+      cudf::type_dispatcher(output.get_column(col)->dtype,
+                            scatter_functor<Filter, block_size, per_thread>{},
+                            *output.get_column(col), *input.get_column(col), 
+                            block_offsets, filter, stream);
+    }
   }
-  
-  // synchronize nvcategory after filtering
-  if (output.dtype == GDF_STRING_CATEGORY) {
-    CUDF_EXPECTS(
-    GDF_SUCCESS ==
-      nvcategory_gather(&output,
-                        static_cast<NVCategory *>(input.dtype_info.category)),
-      "could not set nvcategory");
+  else {
+    output = empty_like(input);
   }
+  
+  CHECK_STREAM(stream);
+
+  gdf_error err = nvcategory_gather_table(input, output);
+  CHECK_STREAM(0);
+  CUDF_EXPECTS(err == GDF_SUCCESS, "Error updating nvcategory in copy_if");
   return output;
 }
 
+/*
+ * @brief Filters a column using a Filter function object
+ * 
+ * @p filter must be a functor or lambda with the following signature:
+ * __device__ bool operator()(gdf_index_type i);
+ * It returns true if element i of @p input should be copied, false otherwise.
+ * 
+ * @note: filter must return false if its input i is out of bounds for any 
+ * memory accesses, therefore it is expected to know the bounds of any 
+ * arrays it uses internally.
+ *
+ * @tparam Filter the filter functor type
+ * @param[in] input The column to filter-copy
+ * @param[in] filter A function object that takes an index and returns a bool
+ * @return The filter-copied result column
+ */
+template <typename Filter>
+gdf_column copy_if(gdf_column const &input, Filter filter,
+                   cudaStream_t stream = 0)
+{
+  // convert column to table
+  gdf_column * cols[1];
+  cols[0] = const_cast<gdf_column*>(&input);
+  const table input_table(cols, 1);
```

Comment:
You can just use the initializer list constructor. 
```suggestion
  const table input_table{{&input}};
```

Reply:
This is gone now.